File: server/ruby/app/infrastructure/mappers/payment_details/payment_status_output_mapper.rb
module Infrastructure
  module Mappers
    module PaymentDetails
      class PaymentStatusOutputMapper
        def self.from_sdk_response(response)
          return nil if response.nil?

          dto = Business::Domain::Payments::PaymentDetails::PaymentStatusOutput.new
          dto.is_authorized = response.is_authorized
          dto.is_cancellable = response.is_cancellable
          dto.is_refundable = response.is_refundable
          dto.status_category = response.status_category
          dto.status_code = response.status_code
          dto.status_code_change_date_time = response.status_code_change_date_time
          dto.errors = map_errors(response.errors)

          dto
        end

        private_class_method def self.map_errors(errors)
          return nil if errors.nil?

          errors.map { |error| Infrastructure::Mappers::PaymentDetails::ApiErrorMapper.from_sdk_response(error) }
        end
      end
    end
  end
end

File: server/python/app/infrastructure/mappers/payment_details/payment_status_output_mapper.py
from typing import Optional, List
from app.application.domain.payments.payment_details.payment_status_output import PaymentStatusOutput
from app.application.domain.payments.payment_details.api_error import APIError
from onlinepayments.sdk.domain.payment_status_output import PaymentStatusOutput as PaymentStatusOutputSdk
from onlinepayments.sdk.domain.api_error import APIError as APIErrorSdk
from app.infrastructure.mappers.payment_details.api_error_mapper import ApiErrorMapper

class PaymentStatusOutputMapper:

    @staticmethod
    def map_from_sdk_response(response: Optional[PaymentStatusOutputSdk]) -> Optional[PaymentStatusOutput]:
        if response is None:
            return None

        return PaymentStatusOutput(
            is_authorized = response.is_authorized,
            is_cancellable = response.is_cancellable,
            is_refundable = response.is_refundable,
            status_category = response.status_category,
            status_code = response.status_code,
            status_code_change_date_time = response.status_code_change_date_time,
            errors = PaymentStatusOutputMapper.map_list(response.errors)
        )

    @staticmethod
    def map_list(errors: Optional[List[APIErrorSdk]]) -> Optional[List[APIError]]:
        if errors is None:
            return None
        return [ApiErrorMapper.map_from_sdk_response(e) for e in errors]